### ISSUE TITLE
Navigation inside decompiled files

### ISSUE THREAD

**Issue by seblj** [seblj/roslyn.nvim#116]
I see that vscode now has support for navigation inside a decompiled file. This is not the case for this plugin. It only has some very simple hover information on some things.

I want to be able to keep navigating in the decompiled files.

Currently, we are falling back to the currently attached solution, and I think that was the case for vscode previously, but it seems they must have changed something to make it better

**Comment by searleser97** [comment]
I am able to navigate through decompiled files, is this still an issue ? or was it just fixed lately ?

**Comment by seblj** [comment]
I tested this yesterday when I opened the issue🤔 do you have a small video of you doing this? I am able to navigate to a decompiled file, but I can't continue navigating

**Comment by nm-marques** [comment]
Can confirm this works fine for me too, same behaviour as vscode, gd works fine inside decompiled files even to other decompiled files. I can record something in a test project when i can.

**Comment by seblj** [comment]
Super weird

**Comment by ashishbinu** [comment]
yes, I have the same issue. It can't go to other definition of the decompiled file.

**Comment by DerrikCreates** [comment]
I'm pretty new to neovim over all. But I've been able to get it to work sometimes. This [commit](https://github.com/DerrikCreates/dotfiles/tree/0b9f37a8760085feef9c1c0fefbaff38257e02e9) is my config ive been using where its sometimes working. 

`Windows 11 23H2`
`NVIM v0.10.1`


The language server files im using came from `Microsoft.CodeAnalysis.LanguageServer.4.8.0-3.23451.1.nupkg` from the manual setup in the readme. 


The key map im using to navigate tot eh definition is this
`keymap.set("n", "gt", "<cmd>Telescope lsp_type_definitions<CR>", opts)`

At first today i tried setting `lock_target = true` and it started working after setting it to true but it might be a red herring. Because when its not working i get this message `[telescope.builtin.lsp_*]: no client attached]`. It only says this when trying to go to a definition in an already decompiled buffer, it still works in buffers that are **not** decompiled.

here is a video of when i got it working. Closing and opening neovim caused it to stop working.
https://github.com/user-attachments/assets/19bfa996-49c9-48f4-b6d2-af4e8fac4281


here is what happens when its not working.
https://github.com/user-attachments/assets/079fea42-ca71-4f08-88c7-199469606abd

I'm very new to neovim, I only started using it as a new years resolution as a pivot from IdeaVim. So its very possible I made a mistake when making [my config](https://github.com/DerrikCreates/dotfiles/tree/0b9f37a8760085feef9c1c0fefbaff38257e02e9). Im not sure what I did to get it working in the first place and im not really sure what to change to try and force these decompiled temp file buffers to attach to the lsp(if this is even the issue).

I'm going to keep messing around to find a solution, I was screen recording during making this work but its a lot of video to look through. If there is any more information i can give to make solving this easier let me know

**Comment by DerrikCreates** [comment]
I figured it out and it might explain why it works for some and not others.

For some reason the initial auto targeting of the solution seems to break going to definitions. `Roslyn target` ing after the initial load will fix it **but only in buffers opened after resetting** the target. Setting the target will not fix already open buffers

My guess some people always set the target without relying on the auto targeting.

`lock_target` does not seem to affect the workaround


**Issue**:

1. open file in solution letting the plugin auto target
2. goto def
3. goto def inside of the decompiled file
4. Fails 

**Workaround**:

1. open file in solution
2. `Roslyn target` and select your solution
3. Close any existing decompiled buffers
4. goto def
5. goto def inside of the decompiled file
6. Works as expected


here is a video of the workaround working
https://github.com/user-attachments/assets/b94f66ce-3642-480f-8137-44328bddec81

**Comment by drzbida** [comment]
that's very weird; i'm not using lock target / CSTarget and yet navigation still works for me in decompiled files

**Comment by seblj** [comment]
Hmm, I tried to do the same things as you did and I am still not able to navigate in decompiled code like that🤔

**Comment by seblj** [comment]
Okay I just tried on my windows VM, and there it seems to work. However, it does not work on macOS for me

**Comment by Sverre-W** [comment]
Related to navigation in decompiled files, when I try to navigate to defintion of a type that is in a different project it will also navigate to the decompiled version of that type. Is this a limitation of the LSP itself? 

Related to this seems that the LSP intellisense uses a the decompiled version to provide intellisense for a type from the referenced project, adding a field for example, I have to recompile the referenced project and restart the language server in order for the diagnostics and autocomplete to pick this up thse changes in the referenced project.

All of the projects are part of a single solution.

**Comment by hieulw** [comment]
> Having the same problem here. This is happening with the new versions. When using old versions of nvim and roslyn it worked well on Linux at least

May I ask which version you use that makes it work on linux ?

**Comment by drzbida** [comment]
I've switched to linux since the last response and I can still navigate across decompiled files with the last 3 server versions. If someone is willing to provide a minimal full nvim config where it doesn't work and the associated roslyn version, I will look into it.

**Comment by benjiwolff** [comment]
When I open an external source file with roslyn nvim, it is located at 
`/private/var/folders/_c/27x2ds_55459nbcfj6px_gyr0000gn/T/MetadataAsSource/cbdbe3ed14be41aa842a9eef0846ed62/DecompilationMetadataAsSourc
eFileProvider/9865c76504b14293be42b2539d48f0d2/FileConfigurationSource.cs`.

When I open it with vscode, it islocated at `~/.dotnet/symbolcache/FileConfigurationSource.cs`.

I am running in the issue as well. Could it be the file location that is handled differently and causes issues?

**Comment by franroa** [comment]
Hi, I have reinstalled my neovim config without changing anything. Now I can navigate inside those files

**Comment by benjiwolff** [comment]
After some extensive debugging of the language server with VSCode and Neovim, I found the following:
* [`PdbFileLocatorService`](https://github.com/dotnet/roslyn/blob/bdf23fd720c03ed09a90359edaa3508c32875d3a/src/Features/Core/Portable/PdbSourceDocument/PdbFileLocatorService.cs#L18) fails to open the `.pdb` symbol file only with neovim. This is because `System.Reflection.PortableExecutable.PEReader` expects the `.pdb` file to be next to the dll (e.g. /usr/local/share/dotnet/shared/Microsoft.AspNetCore.App/9.0.4/Microsoft.Extensions.DependencyInjection.Abstractions.pdb), which is not the case on my macbook.
* When using VSCode, `PdbFileLocatorService` contains a special [`VSCodeSourceLinkService`](https://github.com/dotnet/roslyn/blob/bdf23fd720c03ed09a90359edaa3508c32875d3a/src/VisualStudio/DevKit/Impl/SourceLink/VSCodeSourceLinkService.cs) that performs some fallback.

I will continue to look into this. Just thought I'd give an update, since it took me quite a while to get this far 😅

**Comment by benjiwolff** [comment]
Mystery solved: the language server returns a symbolic link for decompiled files (at least on macOS `/var` instead of `/private/var`). This is expanded by Neovim to `/private/var` in subsequent requests (e.g. `textDocument/definition`). The expanded path is not recognized properly.

Here's a dirty workaround that for which you need to compile the language server yourself https://github.com/dotnet/roslyn/issues/78611.

I guess it's also something that `roslyn.nvim` could fix with a custom `vim.lsp.rpc.PublicClient`.
https://neovim.io/doc/user/lsp.html#vim.lsp.Config

**Comment by franroa** [comment]
Interesting. That explain why now is working on my WSL. I installed wslpath lately

**Comment by franroa** [comment]
For me stop working

**Comment by dssste** [comment]
I am able to navigate inside decompiled files with `roslyn 5.1.0-1.25506.3`.

I am not using this repo anymore and my config is kind of dumb
```lua
	root_dir = function(bufnr, on_dir)
		local bufname = vim.api.nvim_buf_get_name(bufnr)

		if bufname:match("^roslyn%-source%-generated://") then return end

		if bufname:match("/MetadataAsSource/") then
			local client = vim.lsp.get_clients({ name = "roslyn" })[1]
			if client and client.config.root_dir then
				on_dir(client.config.root_dir)
			end
			return
		end

		local root_dir = vim.fs.root(bufnr, function(fname, _)
			return fname:match('%.sln$') ~= nil
		end)
		if not root_dir then
			root_dir = vim.fs.root(bufnr, function(fname, _)
				return fname:match('%.csproj$') ~= nil
			end)
		end
		if root_dir then
			on_dir(root_dir)
		end
	end,
```
(macOS, had the problem of not able to navigate before)
